File: src/ui/common/src/utils/storage.ts
import { IntegrationConfig, Service } from './resources';

export enum StorageType {
  S3 = 's3',
  File = 'file',
  GCS = 'gcs',
}

export const StorageTypeNames = {
  s3: 'AWS S3',
  file: 'Local File System',
  gcs: 'Google Cloud Storage',
};

export type S3Config = {
  region: string;
  bucket: string;
  credentials_path?: string;
  credentials_profile?: string;
  aws_access_key_id?: string;
  aws_secret_access_key?: string;
};

export type FileConfig = {
  directory: string;
};

export type GCSConfig = {
  bucket: string;
  service_account_credentials?: string;
};

export type StorageConfig = {
  type: StorageType;
  s3_config?: S3Config;
  file_config?: FileConfig;
  gcs_config?: GCSConfig;
};

export type MetadataStorageConfig = {
  type: StorageType;
  s3Config?: S3Config;
  fileConfig?: FileConfig;
  gcsConfig?: GCSConfig;
};

export type ServerConfig = {
  aqPath: string;
  retentionJobPeriod: string;
  apiKey: string;
  storageConfig: MetadataStorageConfig;
};

function convertS3IntegrationtoMetadataStorageConfig(
  storage: S3Config
): MetadataStorageConfig {
  return {
    type: StorageType.S3,
    s3Config: {
      region: storage.region,
      bucket: 's3://' + storage.bucket,
    },
  };
}

function convertGCSIntegrationtoMetadataStorageConfig(
  storage: GCSConfig
): MetadataStorageConfig {
  return {
    type: StorageType.GCS,
    gcsConfig: {
      bucket: storage.bucket,
    },
  };
}

export function convertIntegrationConfigToServerConfig(
  storage: IntegrationConfig,
  metadataStorage: ServerConfig,
  service: Service
): ServerConfig {
  let storageConfig;
  switch (service) {
    case 'S3':
      storageConfig = convertS3IntegrationtoMetadataStorageConfig(
        storage as S3Config
      );
      break;
    case 'GCS':
      storageConfig = convertGCSIntegrationtoMetadataStorageConfig(
        storage as GCSConfig
      );
      break;
    default:
      return null;
  }
  return {
    aqPath: metadataStorage.aqPath,
    retentionJobPeriod: metadataStorage.retentionJobPeriod,
    apiKey: metadataStorage.apiKey,
    storageConfig: storageConfig,
  };
}

File: src/golang/lib/models/shared/storage.go
package shared

import (
	"database/sql/driver"


	"github.com/aqueducthq/aqueduct/lib/models/utils"
	"github.com/dropbox/godropbox/errors"
	"github.com/google/uuid"
)

type StorageType string

const (
	S3StorageType   StorageType = "s3"
	FileStorageType StorageType = "file"
	GCSStorageType  StorageType = "gcs"
)

type StorageConfig struct {
	Type       StorageType `yaml:"type" json:"type"`
	S3Config   *S3Config   `yaml:"s3Config" json:"s3_config,omitempty"`
	FileConfig *FileConfig `yaml:"fileConfig" json:"file_config,omitempty"`
	GCSConfig  *GCSConfig  `yaml:"gcsConfig"  json:"gcs_config,omitempty"`
}

type StorageConfigPublic struct {
	Type            StorageType      `json:"type"`
	S3ConfigPublic  *S3ConfigPublic  `json:"s3Config,omitempty"`
	FileConfig      *FileConfig      `json:"fileConfig,omitempty"`
	GCSConfigPublic *GCSConfigPublic `json:"gcsConfig,omitempty"`

	// These fields do not exist on the config file, but are pulled in from the database.
	// Empty fields here mean that the local filesystem is being used as storage.
	IntegrationID   uuid.UUID       `json:"integration_id,omitempty"`
	IntegrationName string          `json:"integration_name,omitempty"`
	ConnectedAt     int64           `json:"connected_at,omitempty"`
	ExecState       *ExecutionState `json:"exec_state,omitempty"`
}

type S3Config struct {
	Region string `yaml:"region" json:"region"`
	Bucket string `yaml:"bucket" json:"bucket"`

	// Use this directory in the bucket as the root. If not set, we default to the root of the bucket.
	// Expected to be santizied into the format "path/to/dir/" (without a leading slash, but with a trailing one).
	RootDir string `yaml:"root_dir" json:"root_dir"`

	CredentialsPath    string `yaml:"credentialsPath" json:"credentials_path"`
	CredentialsProfile string `yaml:"credentialsProfile"  json:"credentials_profile"`
	AWSAccessKeyID     string `yaml:"awsAccessKeyId"  json:"aws_access_key_id"`
	AWSSecretAccessKey string `yaml:"awsSecretAccessKey"  json:"aws_secret_access_key"`
}

type S3ConfigPublic struct {
	Region string `yaml:"region" json:"region"`
	Bucket string `yaml:"bucket" json:"bucket"`

	// Use this directory in the bucket as the root. If not set, we default to the root of the bucket.
	// Expected to be santizied into the format "path/to/dir/" (without a leading slash, but with a trailing one).
	RootDir string `yaml:"root_dir" json:"root_dir"`
}

type FileConfig struct {
	Directory string `yaml:"directory" json:"directory"`
}

type GCSConfig struct {
	Bucket                    string `yaml:"bucket"  json:"bucket"`
	ServiceAccountCredentials string `yaml:"serviceAccountCredentials"  json:"service_account_credentials"`
}

type GCSConfigPublic struct {
	Bucket string `yaml:"bucket"  json:"bucket"`
}

func (s *StorageConfig) Scan(value interface{}) error {
	return utils.ScanJSONB(value, s)
}

func (s *StorageConfig) Value() (driver.Value, error) {
	return utils.ValueJSONB(*s)
}

func (s *StorageConfig) ToPublic() (*StorageConfigPublic, error) {
	storageConfigPublic := &StorageConfigPublic{
		Type: s.Type,
	}

	switch s.Type {
	case FileStorageType:
		storageConfigPublic.FileConfig = s.FileConfig
	case S3StorageType:
		storageConfigPublic.S3ConfigPublic = &S3ConfigPublic{
			Region:  s.S3Config.Region,
			Bucket:  s.S3Config.Bucket,
			RootDir: s.S3Config.RootDir,
		}
	case GCSStorageType:
		storageConfigPublic.GCSConfigPublic = &GCSConfigPublic{
			Bucket: s.GCSConfig.Bucket,
		}
	default:
		return nil, errors.Newf("Unknown storage type. %s", s.Type)
	}

	return storageConfigPublic, nil
}
